iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -0,0 +1,368 @@
+    if (isParallelIterator(iteratorTypes[i])) {
+      parallel_iterators.push_back(i);
+    } else if (isReductionIterator(iteratorTypes[i])) {
+      reduction_iterators.push_back(i);
+    } else {
+      return false;
+    }
+  }
+  if (parallel_iterators.size() != 2 || reduction_iterators.size() != 1) {
+    return false;
+  }
+  // Give the found iterators some idiomatic names.
+  const int MIter = parallel_iterators[0];
+  const int NIter = parallel_iterators[1];
+  const int KIter = reduction_iterators[0];
+  // Check that there are 3 indexing maps.
+  auto indexingMaps = contractionOp.indexing_maps().getValue();
+  if (indexingMaps.size() != 3) {
+    return false;
+  }
+  // Check that the indexing maps have the expected form.
+  const int expectedMapResults[3][2] = {
+      {MIter, KIter}, {NIter, KIter}, {MIter, NIter}};
+  for (int m = 0; m < 3; ++m) {
+    auto map = indexingMaps[m].cast<AffineMapAttr>().getValue();
+    if (map.getNumDims() != 3 || map.getNumResults() != 2) {
+      return false;
+    }
+    for (int r = 0; r < 2; ++r) {
+      int actualMapResult =
+          map.getResults()[r].cast<AffineDimExpr>().getPosition();
+      if (actualMapResult != expectedMapResults[m][r]) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+// Returns true if `contractionOp` is of the form
+//   matrix * transposed_matrix
+// where matrix is a vector<{mSize}x{kSize}xType>, and
+// transposed_matrix is a vector<{nSize}x{kSize}xType>
+static bool isMatrixTimesMatrixTransposedOfGivenShape(
+    vector::ContractionOp contractionOp, int64_t mSize, int64_t kSize,
+    int64_t nSize) {
+  if (!isMatrixTimesMatrixTransposed(contractionOp)) {
+    return false;
+  }
+  VectorType lhsType = contractionOp.lhs().getType().cast<VectorType>();
+  VectorType rhsType = contractionOp.rhs().getType().cast<VectorType>();
+  auto lhsShape = lhsType.getShape();
+  auto rhsShape = rhsType.getShape();
+  if (lhsShape[0] != mSize || lhsShape[1] != kSize || rhsShape[0] != nSize ||
+      rhsShape[1] != kSize) {
+    return false;
+  }
+  return true;
+}
+
+// Checks that the Value `extResult` is defined by an arith::ExtSIOp promoting
+// from `extSrcType` to `extDstType`, and returns the input of the ExtSIOp.
+static Value getExtSIInput(Type extSrcType, Type extDstType, Value extResult) {
+  auto extSIOp = extResult.getDefiningOp<arith::ExtSIOp>();
+  if (!extSIOp) {
+    return nullptr;
+  }
+  Value extInput = extSIOp.getIn();
+  if (extInput.getType().cast<VectorType>().getElementType() != extSrcType) {
+    return nullptr;
+  }
+  return extInput;
+}
+
+// Helper to create a 1D, contiguous slice of a 1D vector.
+static Value extract1DSlice(PatternRewriter &rewriter, Location loc,
+                            VectorType dstVecType, Value input, int position) {
+  assert(input.getType().cast<VectorType>().getRank() == 1);
+  assert(dstVecType.getRank() == 1);
+  std::array<int64_t, 1> offsets{position};
+  std::array<int64_t, 1> strides{1};
```

Comment:
Actually this created a dangling reference:

```c++
ArrayRef<int64_t> offsets{position};
```

because `{position}` is a temporary that immediately goes out of scope. And I didn't see a compiler warning (Clang 11).

I only noticed because tests failed at runtime with uninitialized values:

```
iree/compiler/Codegen/LLVMCPU/test/vector_contract_custom_kernels.mlir:9:10: error: 'vector.extract_strided_slice' op expected offsets dimension 0 to be confined to [0, 32)
  %res = vector.contract {
         ^
iree/compiler/Codegen/LLVMCPU/test/vector_contract_custom_kernels.mlir:9:10: note: see current operation: %4 = "vector.extract_strided_slice"(%1) {offsets = [115461696], sizes = [16], strides = [1]} : (vector<32xi8>) -> vector<16xi8>
```

This exemplifies that the optimizations that we pepper around MLIR C++ code are not only pointless, they are also: unsafe and wasting developer time.